cmd/internal/planetscale_edge_database.go

```diff
@@ -337,56 +374,76 @@ func (p PlanetScaleEdgeDatabase) getLatestCursorPosition(ctx context.Context, sh
 	ctx, cancel := context.WithTimeout(ctx, timeout)
 	defer cancel()
 	var (
-		err    error
-		client psdbconnect.ConnectClient
+		err          error
+		vtgateClient vtgateservice.VitessClient
 	)
 
-	if p.clientFn == nil {
-		conn, err := grpcclient.Dial(ctx, ps.Host,
-			clientoptions.WithDefaultTLSConfig(),
-			clientoptions.WithCompression(true),
-			clientoptions.WithConnectionPool(1),
-			clientoptions.WithExtraCallOption(
-				auth.NewBasicAuth(ps.Username, ps.Password).CallOption(),
-			),
-		)
-		if err != nil {
-			return "", err
-		}
+	vtgateClient, conn, err := p.initializeVTGateClient(ctx, ps)
+	if err != nil {
+		return "", err
+	}
+	if conn != nil {
 		defer conn.Close()
-		client = psdbconnect.NewConnectClient(conn)
-	} else {
-		client, err = p.clientFn(ctx, ps)
-		if err != nil {
-			return "", err
-		}
 	}
 
-	sReq := &psdbconnect.SyncRequest{
-		TableName: s.Name,
-		Cursor: &psdbconnect.TableCursor{
-			Shard:    shard,
-			Keyspace: keyspace,
-			Position: "current",
-		},
-		TabletType: tabletType,
-		Cells:      []string{"planetscale_operator_default"},
-	}
+	vtgateReq := buildVStreamRequest(tabletType, s.Name, shard, keyspace, "current")
+	vtgateCursor, vtgateErr := vtgateClient.VStream(ctx, vtgateReq)
 
-	c, err := client.Sync(ctx, sReq)
-	if err != nil {
+	if vtgateErr != nil {
 		return "", nil
 	}
 
 	for {
-		res, err := c.Recv()
+		res, err := vtgateCursor.Recv()
 		if err != nil {
 			return "", err
 		}
 
-		if res.Cursor != nil {
-			return res.Cursor.Position, nil
+		if res.Events != nil {
+			for _, event := range res.Events {
+				if event.Type == binlogdata.VEventType_VGTID {
+					gtid := event.Vgtid.ShardGtids[0].Gtid
+					return gtid, nil
+				}
+			}
+			return "", errors.New("unable to find VEvent of VGTID type to use as stop cursor")
+		}
+	}
+}
+
+func toTopoTabletType(tabletType psdbconnect.TabletType) topodata.TabletType {
+	if tabletType == psdbconnect.TabletType_replica {
+		return topodata.TabletType_REPLICA
+	} else if tabletType == psdbconnect.TabletType_batch {
+		return topodata.TabletType_RDONLY
+	} else if tabletType == psdbconnect.TabletType_primary {
+		return topodata.TabletType_PRIMARY
+	}
+
+	// Fall back to replica
+	return topodata.TabletType_REPLICA
+}
+
+func (p PlanetScaleEdgeDatabase) initializeVTGateClient(ctx context.Context, ps PlanetScaleSource) (vtgateservice.VitessClient, grpcclient.ConnPool, error) {
+	if p.vtgateClientFn == nil {
```

Comment:
do we need to support both the `p.vtgateClientFn == nil` path as well as the `!= nil` path?